### PR TITLE
(test) O3-2216: Add tests for `id-field.component.tsx`

### PR DESCRIPTION
**Pull request by ayush-AI**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

In this PR i have added test for id-field component. 

## Screenshots

Screenshot 2023-06-30 at 11 51 16 PM

## Related Issue

https://issues.openmrs.org/browse/O3-2216

## Other

*None.*

### CODE REVIEW

**Comment by ayush-AI** [comment]
cc/ @jayasanka-sack @anjula-sack @Piumal1999

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+import { PatientRegistrationContext } from '../../patient-registration-context';
+import { useConfig } from '@openmrs/esm-framework';
+
+jest.mock('@openmrs/esm-framework', () => ({
+  ...jest.requireActual('@openmrs/esm-framework'),
+  useConfig: jest.fn(),
+}));
+
+describe('Identifiers', () => {
+  const mockedIdentifierTypes = [
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:15]: You can use [\_\_mocks\_\_ directory](https://github.com/openmrs/openmrs-esm-patient-management/tree/main/__mocks__)  to store these kind of mock objects.

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        uuid: '01af8526-cea4-4175-aa90-340acb411771',
+        name: 'Generator 2 for OpenMRS ID',
+        autoGenerationOption: {
+          manualEntryEnabled: true,
+          automaticGenerationEnabled: true,
+        },
+      },
+    ],
+    autoGenerationSource: null,
+  };
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:112]:
```suggestion
  };

```

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+              <Identifiers />
+            </PatientRegistrationContext.Provider>
+          </Form>
+        </Formik>
+      </ResourcesContext.Provider>,
+    );
+
+    expect(screen.getByText('Identifiers')).toBeInTheDocument();
+    expect(screen.getByText('Configure')).toBeInTheDocument();
+    expect(screen.getByText('Configure')).toBeEnabled();
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:163]: If this is related to a Button, we can use getByRole('button', { name: 'Configure' }). We need to make sure that we are querying the correct element.

(Use getByRole() whenever it is possible.)

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+          </Form>
+        </Formik>
+      </ResourcesContext.Provider>,
+    );
+
+    const configureButton = screen.getByText('Configure');
+    fireEvent.click(configureButton);
+
+    expect(screen.getByRole('button', { name: 'Close overlay' })).toBeInTheDocument();
+  });
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:191]:
```suggestion
  });
  
```

**Review by Piumal1999 — CHANGES_REQUESTED** [review]
Good job @ayush-AI. I requested some minor changes.

Btw, there's a formatting pattern called [Arrange-Act-Assert](https://robertmarshall.dev/blog/arrange-act-and-assert-pattern-the-three-as-of-unit-testing/), which is used to improve the readability of tests. You can follow it when arranging the steps of the tests (if it suits our test cases).

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      </ResourcesContext.Provider>,
+    );
+
+    const configureButton = screen.getByRole('button', { name: 'Configure' });
+    fireEvent.click(configureButton);
+
+    expect(screen.getByRole('button', { name: 'Close overlay' })).toBeInTheDocument();
+  });
+
+  it('should close identifier selection overlay when "close overlay" button is clicked', () => {
```
Comment (jayasanka-sack) [inline packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:136]: I think this test is not related to the id-field component. It is something `ui-components/overlay` is responsible for.

Comment (ayush-AI) [reply packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:136]: Yes @jayasanka-sack you are right. This is coming from the overlay component. `id-field.component.tsx` is using `identifier-selection-overlay.tsx` as child component which is wrapper implementation of `ui-components/overlay`. But there is one issue when I will test this behaviour in test for `ui-components/overlay` then the ui changes will not happen. for example the close button will not close the overlay component it will just call the onClose function. So in this case I am not sure what to do ? If I am testing this in parent component the ui change is happening. So where should I place this test?
cc/ @Piumal1999 @anjula-sack

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    currentSession: {
+      authenticated: true,
+      sessionId: 'JSESSION',
+      currentProvider: { uuid: 'provider-uuid', identifier: 'PRO-123' },
+    },
+    relationshipTypes: [],
+    identifierTypes: [...mockedIdentifierTypes],
+  } as Resources;
+
+  const openmrsID = {
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:27]: Can we move this one also to the mocks file?

Comment (ayush-AI) [reply packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx]: done @Piumal1999

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    currentSession: {
+      authenticated: true,
+      sessionId: 'JSESSION',
+      currentProvider: { uuid: 'provider-uuid', identifier: 'PRO-123' },
+    },
+    relationshipTypes: [],
+    identifierTypes: [...mockedIdentifierTypes],
+  } as Resources;
+
+  (useConfig as jest.Mock).mockImplementation(() => ({
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:28]: Instead of this, you can mock the implementation within the above jest.mock() function it self. Refer to [this](https://github.com/openmrs/openmrs-esm-patient-chart/blob/5d2baf9f1ff021ff85ad5a9468237f4a794e3d45/packages/esm-patient-programs-app/src/programs/programs-overview.test.tsx#L20)

File: packages/esm-patient-registration-app/src/patient-registration/field/__mocks__/identifiers.mock.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+export const openmrsID = {
```
[the remark this answers is not in the document]
Comment (ayush-AI) [reply packages/esm-patient-registration-app/src/patient-registration/field/__mocks__/identifiers.mock.ts:1]: I tried using it but the test started failing. I think we should keep both of them.

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+      <ResourcesContext.Provider value={mockResourcesContextValue}>
+        <Formik initialValues={{}} onSubmit={null}>
+          <Form>
+            <PatientRegistrationContext.Provider
+              value={{
+                setFieldValue: jest.fn(),
+                initialFormValues: { identifiers: { openmrsID } },
+                setInitialFormValues: jest.fn(),
+                values: {
+                  identifiers: { openmrsID },
```
Comment (ayush-AI) [inline packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:66]: when I am changing this openmrsID. then the tests start to fail.

> TypeError: Cannot read properties of null (reading 'identifierTypeUuid')

      24 |   const { isOffline, values, setFieldValue } = useContext(PatientRegistrationContext);
      25 |   const identifierType = useMemo(
    > 26 |     () => identifierTypes.find((identifierType) => identifierType.uuid === patientIdentifier.identifierTypeUuid),
         |                                                                                              ^
      27 |     [patientIdentifier, identifierTypes],
      28 |   );
      29 |   const { autoGeneration, initialValue, identifierValue, identifierName, required } = patientIdentifier;

**Review by Piumal1999 — DISMISSED** [review]

File: packages/esm-patient-registration-app/src/patient-registration/field/__mocks__/identifiers.mock.ts (hunk 8 above)
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/field/__mocks__/identifiers.mock.ts:1]: We can remove this file

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
```
[the remark this answers is not in the document]
Comment (Piumal1999) [reply packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:63]: The requested change is not added. Could you please recheck? @ayush-AI

[the remark this answers is not in the document]
Comment (ayush-AI) [reply packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.test.tsx:63]: It is not working like the way you are suggesting. The test is failing because of syntax error

**Comment by jayasanka-sack** [comment]
@ayush-AI could you please check @Piumal1999 's comments?

**Comment by ayush-AI** [comment]
> @ayush-AI could you please check @Piumal1999 's comments?
I checked it and it is not working as piumal suggested to try. That is why I have to keep that file.

cc:/ @Piumal1999

**Comment by denniskigen** [comment]
Can we move this along @ayush-AI @Piumal1999?

**Comment by ayush-AI** [comment]
yup we can @denniskigen. If other thing seems okay to you.

**Review by denniskigen — APPROVED** [review]
